[DOCS] Add PIT to search after docs (#61593)
elastic/elasticsearch: docs/reference/search/search-your-data/paginate-search-results.asciidoc

preserve the current index state for a specified period.

IMPORTANT: We don't recommend creating PITs for each user request. Each PIT
requires additional disk space and file handles. If you want to paginate search
results for user requests, skip this step and use `search_after` without a PIT.

[source,console]
----
POST /my-index-000001/_pit?keep_alive=1m
----
// TEST[setup:my_index]

The API returns a PIT ID.

[source,console-result]
----
{
  "id": "46ToAwMDaWR4BXV1aWQxAgZub2RlXzEAAAAAAAAAAAEBYQNpZHkFdXVpZDIrBm5vZGVfMwAAAAAAAAAAKgFjA2lkeQV1dWlkMioGbm9kZV8yAAAAAAAAAAAMAWICBXV1aWQyAAAFdXVpZDEAAQltYXRjaF9hbGw_gAAAAA=="
}
----
// TESTRESPONSE[s/"id": "46ToAwMDaWR4BXV1aWQxAgZub2RlXzEAAAAAAAAAAAEBYQNpZHkFdXVpZDIrBm5vZGVfMwAAAAAAAAAAKgFjA2lkeQV1dWlkMioGbm9kZV8yAAAAAAAAAAAMAWICBXV1aWQyAAAFdXVpZDEAAQltYXRjaF9hbGw_gAAAAA=="/"id": $body.id/]

To get the first page of results, submit a search request with a `sort`
argument. If using a PIT, include the PIT ID as the `pit.id` argument.

IMPORTANT: We recommend you include a tiebreaker field in your `sort`. This
tiebreaker field should contain a unique value for each document. If you don't

> include a tiebreaker field, your paged results could miss or duplicate hits.

Author: For reviewers:

The original admon suggested using the `set` ingest processor to create a tiebreaker field containing the doc's `_id` value. However, you can't use ingest processors to access autogen docs IDs.

Is it worth explicitly mentioning enabling `indices.id_field_data.enabled` and sorting on `_id` here?

Reviewer: No this should be avoided and in 8.0 this will probably be forbidden. We have a follow up for PIT to add a simple tiebreaker in https://github.com/elastic/elasticsearch/issues/56828. We want to have it soon so that users don't need to worry about creating a doc value field for this purpose. That said, I think that the documentation should be updated when we have the new feature but the new shape looks great. I think we should proceed with this update to the docs and add the virtual tiebreaker part when it's ready.

Author: Sounds good. Thanks for the context @jimczi.

<<index-max-result-window,`index.max_result_window`>> index setting. If you need
to page through more than 10,000 hits, use the <<search-after,`search_after`>>
parameter instead.

WARNING: {es} uses Lucene's internal doc IDs as tie-breakers. These internal doc
IDs can be completely different across replicas of the same data. When paging
search hits, you might occasionally see that documents with the same sort values
are not ordered consistently.

[discrete]
[[search-after]]
=== Search after

You can use the `search_after` parameter to retrieve the next page of hits
using a set of <<sort-search-results,sort values>> from the previous page.

Using `search_after` requires multiple search requests with the same `query` and
`sort` values. If a <<near-real-time,refresh>> occurs between these requests,
the order of your results may change, causing inconsistent results across pages. To
prevent this, you can create a <<point-in-time-api,point in time (PIT)>> to
preserve the current index state over your searches.

IMPORTANT: We don't recommend creating PITs for each user request. Each PIT
requires additional disk space and file handles. If you want to paginate search

> results for user requests, use `search_after` without a PIT.

Reviewer: That's not entirely true. I know we had such recommendations for `search_after` but I think we should remove it now. The goal of the PIT is to provide a way to share the readers internally so that's misleading imo.

Author: Good to know. Removed with df7ca438f5ecda1ab49f40a84ec2716f5280edcf. Thanks for the correction!

To get the next page of results, rerun the previous search using the last hit's
sort values as the `search_after` argument. If using a PIT, use the latest PIT
ID in the `pit.id` parameter. The search's `query` and `sort` arguments must
remain unchanged. If provided, the `from` argument must be `0` (default) or `-1`.

[source,console]
----
GET /my-index-000001/_search
{
  "size": 10000,
  "query": {
    "match" : {
      "user.id" : "elkbee"
    }
  },
  "pit": {
	    "id":  "46ToAwEPbXktaW5kZXgtMDAwMDAxFnVzaTVuenpUVGQ2TFNheUxVUG5LVVEAFldicVdzOFFtVHZTZDFoWWowTGkwS0EAAAAAAAAAAAQURzZzcUszUUJ5U1NMX3Jyak5ET0wBFnVzaTVuenpUVGQ2TFNheUxVUG5LVVEAAA==", <1>
	    "keep_alive": "1m"
  },
  "sort": [
    {"@timestamp": "asc"},
    {"tie_breaker_id": "asc"}
  ],
  "search_after": [                                <2>
    4098435132000,
    "FaslK3QBySSL_rrj9zM5"
  ]
}
----
// TEST[catch:missing]

<1> PIT ID returned by the previous search.
<2> Sort values from the previous search's last hit.

You can repeat this process to get additional pages of results. If using a PIT,
you can extend the PIT's retention period using the
`keep_alive` parameter of each search request.

> When you're finished, we recommend deleting your PIT.

Reviewer: ```suggestion
When you're finished, you should deleting your PIT.
```
